flow/record/adapter/sqlite.py

```diff
@@ -0,0 +1,246 @@
+        ).fetchall()
+
+        fields = []
+        fnames = []
+        fname_to_type = {}
+        for idx, row in enumerate(schema):
+            ftype, fname = row
+            fname = sanitized_name(fname)
+            ftype = SQLITE_FIELD_MAP.get(ftype, "string")
+            fname_to_type[fname] = ftype
+            if fname not in RESERVED_FIELDS:
+                fields.append((ftype, fname))
+            fnames.append(fname)
+
+        descriptor_cls = RecordDescriptor(table_name, fields)
+        cursor = self.con.execute(f'SELECT * FROM "{table_name_org}"')
+        while True:
+            rows = cursor.fetchmany(self.batch_size)
+            if not rows:
+                break
+            for row in rows:
+                row = list(row)
+                # A SQLite database could contain values not matching it's type (non STRICT mode)
+                # So try to clean them up where we can.
+                for idx, value in enumerate(row):
+                    fname = fnames[idx]
+                    ftype = fname_to_type[fname]
+                    if ftype == "varint" and value == "":
+                        row[idx] = None
+                    elif ftype == "bytes":
+                        if value == 0:
+                            row[idx] = None
+                        elif isinstance(value, str):
+                            row[idx] = value.encode("utf-8")
+                yield descriptor_cls.init_from_dict(dict(zip(fnames, row)))
+
+    def __iter__(self):
+        """Iterate over all tables in the database and yield records"""
+        for table_name in self.table_names():
+            logging.debug("Reading table: %s", table_name)
+            yield from self.read_table(table_name)
+
+
+class SqliteWriter(AbstractWriter):
+    def __init__(self, path, **kwargs):
+        self.descriptors_seen = set()
+        self.con = sqlite3.connect(path)
+        self.con.isolation_level = None
+        self.count = 0
+        self.batch_size = 1000
+        self.con.execute("PRAGMA journal_mode='wal';")
+
+    def write(self, record: Record) -> None:
+        """Write a record to the database"""
+        desc = record._desc
+        if desc not in self.descriptors_seen:
+            self.descriptors_seen.add(desc)
+            create_descriptor_table(self.con, desc)
+            update_descriptor_columns(self.con, desc)
+
+        db_insert_record(self.con, record)
+        self.count += 1
+        # Commit every batch_size records
+        if self.count % self.batch_size == 0:
+            self.tx_cycle()
+
+    def tx_begin(self):
+        if not self.con.in_transaction:
+            self.con.execute("BEGIN;")
+
+    def tx_commit(self):
+        if self.con.in_transaction:
+            self.con.execute("COMMIT;")
+
+    def tx_cycle(self):
+        self.tx_commit()
+        self.tx_begin()
+
+    def flush(self):
+        if self.con:
+            self.con.commit()
+
+    def close(self):
+        if self.con:
```

Comment:
Will this error if we fail to "connect" to a database? Invalid path for example.

Reply:
yes, you are right. Fixed by always setting `self.con = None` first.